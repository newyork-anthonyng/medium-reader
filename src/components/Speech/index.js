/* global SpeechSynthesisUtterance */

/*
 * https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesis
 */
import React, { Component } from 'react'
import T from 'prop-types'
import speechChunker from '../../utility/speechUtteranceChunker'

class Speech extends Component {
  constructor () {
    super()

    this.state = {
      voices: [],
      selectedVoice: ''
    }

    this.speak = this.speak.bind(this)
    this.pause = this.pause.bind(this)
    this.play = this.play.bind(this)
    this.handleVoiceChange = this.handleVoiceChange.bind(this)
  }

  componentDidMount () {
    /*
     * The SpeechSynthesis voices gets loaded async
     * https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesis/onvoiceschanged
     */
    window.speechSynthesis.onvoiceschanged = () => {
      const voices = window.speechSynthesis.getVoices()
      /*
       * Store all voices in an Object, with each key being the voices name
       */
      this.voicesMap = {}
      voices.forEach((voice) => {
        this.voicesMap[voice.name] = voice
      })

      this.setState({
        voices,
        selectedVoice: 'Google US English'
      })
    }
  }

  speak (text) {
    window.speechSynthesis.cancel()

    /*
     * Create a new utterance with the selected voice, and at 0.8 playback speed
     */
    const msg = new SpeechSynthesisUtterance(text)
    msg.voice = this.voicesMap[this.state.selectedVoice]
    msg.rate = 0.8

    speechChunker(msg, {
      chunkLength: 120
    }, this.props.onEnd)
  }

  play (text) {
    if (window.speechSynthesis.paused) {
      window.speechSynthesis.resume()
    } else {
      this.speak(text)
    }
  }

  pause () {
    window.speechSynthesis.pause()
  }

  handleVoiceChange (e) {
    this.setState({
      selectedVoice: e.target.value
    })
  }

  render () {
    return (
      <div>
        <select value={this.state.selectedVoice} onChange={this.handleVoiceChange}>
          {this.state.voices.map((voice) =>
            <option key={voice.name} value={voice.name}>{voice.name}</option>
          )}
        </select>
      </div>
    )
  }
}

Speech.propTypes = {
  onEnd: T.func
}

export default Speech
