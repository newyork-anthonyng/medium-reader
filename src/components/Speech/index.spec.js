/* global jest, describe, it, expect, beforeEach */

import React from 'react'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import Speech from './'
import speechChunker from '../../utility/speechUtteranceChunker'

jest.mock('../../utility/speechUtteranceChunker')

beforeEach(() => {
  speechChunker.mockClear()

  window.speechSynthesis = {
    speak: jest.fn(),
    resume: jest.fn(),
    pause: jest.fn(),
    cancel: jest.fn(),
    getVoices: jest.fn(() => {
      return [
        { name: 'Voice 1' },
        { name: 'Voice 2' },
        { name: 'Voice 3' },
        { name: 'Google US English' }
      ]
    })
  }
  window.SpeechSynthesisUtterance = jest.fn((text) => ({ text }))
})

const defaultProps = {
  onEnd: () => {}
}

describe('Speech', () => {
  describe('when mounting', () => {
    it('should create voicesMap when component mounts', () => {
      const wrapper = mount(
        <Speech {...defaultProps} />
      )
      window.speechSynthesis.onvoiceschanged()

      expect(wrapper.instance().voicesMap).toMatchSnapshot()
    })

    it('should set default voice when SpeechSynthesis voices loads', () => {
      const wrapper = mount(
        <Speech {...defaultProps} />
      )
      window.speechSynthesis.onvoiceschanged()

      expect(wrapper.state()).toMatchSnapshot()
    })
  })

  it('should render <select> and <options> correctly', () => {
    const wrapper = mount(
      <Speech {...defaultProps} />
    )
    window.speechSynthesis.onvoiceschanged()

    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  it('should add new Utterance and clear out old ones when speaking', () => {
    const cb = jest.fn()
    const wrapper = mount(
      <Speech
        {...defaultProps}
        onEnd={cb}
      />
    )
    window.speechSynthesis.onvoiceschanged()
    const speech = 'Hello World'

    wrapper.instance().speak(speech)

    expect(window.speechSynthesis.cancel).toHaveBeenCalledTimes(1)
    expect(speechChunker).toHaveBeenCalledTimes(1)
    expect(speechChunker.mock.calls[0]).toMatchSnapshot()
    expect(speechChunker.mock.calls[0][2]).toEqual(cb)
  })

  describe('when playing', () => {
    it('should resume if it was previously paused', () => {
      const wrapper = mount(
        <Speech {...defaultProps} />
      )
      window.speechSynthesis.onvoiceschanged()
      window.speechSynthesis.paused = true

      wrapper.instance().play()

      expect(window.speechSynthesis.resume).toHaveBeenCalledTimes(1)
      expect(speechChunker).toHaveBeenCalledTimes(0)
    })

    it('should create new Utterance if it was not previously paused', () => {
      const wrapper = mount(
        <Speech {...defaultProps} />
      )
      window.speechSynthesis.onvoiceschanged()
      window.speechSynthesis.paused = false

      wrapper.instance().play()

      expect(window.speechSynthesis.resume).toHaveBeenCalledTimes(0)
      expect(speechChunker).toHaveBeenCalledTimes(1)
    })
  })

  it('should pause', () => {
    const wrapper = mount(
      <Speech {...defaultProps} />
    )
    window.speechSynthesis.onvoiceschanged()

    wrapper.instance().pause()

    expect(window.speechSynthesis.pause).toHaveBeenCalledTimes(1)
  })

  it('should update <select> when changing options', () => {
    const wrapper = mount(
      <Speech {...defaultProps} />
    )
    window.speechSynthesis.onvoiceschanged()
    const selectTag = wrapper.find('select')

    selectTag.simulate('change', { target: { value: 'Batman' } })
    expect(wrapper.state().selectedVoice).toMatchSnapshot()
  })
})
