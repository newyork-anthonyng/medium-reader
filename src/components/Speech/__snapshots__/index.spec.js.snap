// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Speech should add new Utterance and clear out old ones when speaking 1`] = `
Array [
  Object {
    "rate": 0.8,
    "text": "Hello World",
    "voice": Object {
      "name": "Google US English",
    },
  },
  Object {
    "chunkLength": 120,
  },
  [Function],
]
`;

exports[`Speech should render <select> and <options> correctly 1`] = `
<Speech
  onEnd={[Function]}
>
  <div>
    <select
      onChange={[Function]}
      value="Google US English"
    >
      <option
        value="Voice 1"
      >
        Voice 1
      </option>
      <option
        value="Voice 2"
      >
        Voice 2
      </option>
      <option
        value="Voice 3"
      >
        Voice 3
      </option>
      <option
        value="Google US English"
      >
        Google US English
      </option>
    </select>
  </div>
</Speech>
`;

exports[`Speech should update <select> when changing options 1`] = `"Batman"`;

exports[`Speech when mounting should create voicesMap when component mounts 1`] = `
Object {
  "Google US English": Object {
    "name": "Google US English",
  },
  "Voice 1": Object {
    "name": "Voice 1",
  },
  "Voice 2": Object {
    "name": "Voice 2",
  },
  "Voice 3": Object {
    "name": "Voice 3",
  },
}
`;

exports[`Speech when mounting should set default voice when SpeechSynthesis voices loads 1`] = `
Object {
  "selectedVoice": "Google US English",
  "voices": Array [
    Object {
      "name": "Voice 1",
    },
    Object {
      "name": "Voice 2",
    },
    Object {
      "name": "Voice 3",
    },
    Object {
      "name": "Google US English",
    },
  ],
}
`;
